File: go/resistor-color-trio/resistor_color_trio.go
package resistorcolortrio

import (
	"fmt"
	"math"
	"strconv"
)

// Label describes the resistance value given the colors of a resistor.
// The label is a string with a resistance value with an unit appended
// (e.g. "33 ohms", "470 kiloohms").
func Label(colors []string) string {

	// value determination
	resistorValue := 0
	for index, color := range colors {
		if index > 2 {
			break
		} else if index == 0 {
			resistorValue += resistorColor[color]
		} else if index == 2 {
			resistorValue = resistorValue * int(math.Pow10(resistorColor[color]))
		} else {
			resistorValue = (resistorValue * 10) + resistorColor[color]
		}
	}

	// string conversion
	resistorValueString := ""
	if resistorValue%int(math.Pow10(9)) == 0 && resistorValue != 0 {
		resistorValueString = strconv.Itoa(resistorValue/int(math.Pow10(9))) + " gigaohms"
		fmt.Println(resistorValueString)
	} else if resistorValue%int(math.Pow10(6)) == 0 && resistorValue != 0 {
		resistorValueString = strconv.Itoa(resistorValue/int(math.Pow10(6))) + " megaohms"
		fmt.Println(resistorValueString)
	} else if resistorValue%int(math.Pow10(3)) == 0 && resistorValue != 0 {
		resistorValueString = strconv.Itoa(resistorValue/int(math.Pow10(3))) + " kiloohms"
		fmt.Println(resistorValueString)
	} else {
		resistorValueString = strconv.Itoa(resistorValue) + " ohms"
		fmt.Println(resistorValueString)
	}

	return resistorValueString
}

var resistorColor = map[string]int{
	"black":  0,
	"brown":  1,
	"red":    2,
	"orange": 3,
	"yellow": 4,
	"green":  5,
	"blue":   6,
	"violet": 7,
	"grey":   8,
	"white":  9,
}

File: typescript/resistor-color-trio/resistor-color-trio.ts
export const COLORS = [
  'black',
  'brown',
  'red',
  'orange',
  'yellow',
  'green',
  'blue',
  'violet',
  'grey',
  'white',
] as const;

const units = ["", "kilo", "mega", "giga"];

export function decodedResistorValue(bands: string[]): string {
  // throw new Error('Remove this statement and implement this function')
  let index = -1
  let counter = 0
  let value = 0

  // find the value
  for (let color of bands) {
    let index = -1
    index = COLORS.findIndex(c => c === color.toLowerCase())
    counter++;
    if (index === -1){
      throw new Error('invalid color: ' + color)
    }
    if (counter === 1) {
      value = index*10
    }else if (counter === 2) {
      value += index
    }else if (counter === 3) {
      value = value * (10 ** index)
    }
  }


  // format the number in string
  return formatNumber(value)
}

function formatNumber (num : number): string {
  let unitIndex = 0;

    while (num >= 1000 && unitIndex < units.length - 1) {
        num /= 1000;
        unitIndex++;
    }

    return `${num} ${units[unitIndex]}ohms`;
}


